src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -218,93 +221,202 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
+                            return (int)(offset + (BitOperations.TrailingZeroCount(matches) / sizeof(char)));
+                        }
+                    }
 
-                    if (pCh[0] == value)
-                        goto Found;
+                    lengthToExamine = GetCharVector256SpanLength(offset, length);
+                    if (lengthToExamine > offset)
+                    {
+                        Vector256<ushort> values = Vector256.Create(value);
+                        do
+                        {
+                            Vector256<ushort> search = LoadVector256(ref searchSpace, offset);
+                            int matches = Avx2.MoveMask(Avx2.CompareEqual(values, search).AsByte());
+                            // Note that MoveMask has converted the equal vector elements into a set of bit flags,
+                            // So the bit position in 'matches' corresponds to the element offset.
+                            if (matches == 0)
+                            {
+                                // Zero flags set so no matches
+                                offset += Vector256<ushort>.Count;
+                                continue;
+                            }
+
+                            // Find bitflag offset of first match and add to current offset, 
+                            // flags are in bytes so divide for chars
+                            return (int)(offset + (BitOperations.TrailingZeroCount(matches) / sizeof(char)));
+                        } while (lengthToExamine > offset);
+                    }
 
-                    pCh++;
+                    lengthToExamine = GetCharVector128SpanLength(offset, length);
+                    if (lengthToExamine > offset)
+                    {
+                        Vector128<ushort> values = Vector128.Create(value);
+                        Vector128<ushort> search = LoadVector128(ref searchSpace, offset);
+
+                        // Same method as above
+                        int matches = Sse2.MoveMask(Sse2.CompareEqual(values, search).AsByte());
+                        if (matches == 0)
+                        {
+                            // Zero flags set so no matches
+                            offset += Vector128<ushort>.Count;
+                        }
+                        else
+                        {
+                            // Find bitflag offset of first match and add to current offset, 
+                            // flags are in bytes so divide for chars
+                            return (int)(offset + (BitOperations.TrailingZeroCount(matches) / sizeof(char)));
+                        }
+                    }
+
+                    if (offset < length)
+                    {
+                        lengthToExamine = length - offset;
+                        goto SequentialScan;
+                    }
                 }
+            }
+            else if (Sse2.IsSupported)
```

Comment:
Just a thought. I wonder if it would be worthwhile to create an internal `PlatformVector` abstraction that handles the `256-bit` vs `128-bit` code paths for these operations...

The instructions exposed between the two are generally almost identical and the differences between the code-paths can generally be reduced to incrementing 16 or 32 bytes at a time.

It might make the usages simpler overall.

Reply:
The 256-path uses a loop of 256-bit and then an if for a single 128-bit length
The 128-bit path uses a loop of 128-bit; so they are different

Reply:
It needs to do this to maintain alignment when scanning an unknown length to not fault; e.g. when used by `wcslen(char* ptr)` which calls `SpanHelpers.IndexOf(ref *ptr, '\0', int.MaxValue)`

Reply:
> when scanning an unknown length to not fault;

Do you mean to not read across a cache-line or page boundary? Why not just handle that as part of the "leading elements" to become aligned?

Reply:
Does, but uses Sse as part of it. So it could do 7 (char) => 8 (Sse) => 16 (Avx2) if it was unaligned by  from Avx by 15 chars

Reply:
Right, but couldn't that just be:

```csharp
if (Avx2.IsSupported)
{
    // specialized adjustment
}

// Shared logic
```

NOTE: Definitely not saying it needs to be done in this PR; just an idea to possibly improve these and simplify things later